csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,552 @@ bool predicateAtEnd(ForLoop* loop) {
   return true;
 }
 
+// Implementation of:
+//   Val* proveLinearAndGetStride(
+//       const ValGraph& id_graph,
+//       const ValGroup& linear_g,
+//       const ValGroups& domain);
+//
+// The idea is to propagate from linear_g to domain, keep track of how linear_g
+// lives in domain. For example, let's consider the following two schedules
+// (same example as the NVFuserTest.ProveLinearAndGetStride test):
```

Comment:
Perhaps it would help to define "linear" in this context first, in terms of `IterDomain`s instead of indices. I believe it's similar in spirit to stride compatibility if domain represents the size of a strided tensor. In that case, `linear_g` being linear in `domain` would correspond to the condition that there exists a "view" of the strided tensor such that `linear_g` is one of the new strided axes. That is, it's possible to represent the index for `linear_g` as a combination of the indices for `domain`. Is this correct?

Reply:
> Perhaps it would help to define "linear" in this context first, in terms of `IterDomain`s instead of indices. I believe it's similar in spirit to stride compatibility if domain represents the size of a strided tensor. In that case, `linear_g` being linear in `domain` would correspond to the condition that there exists a "view" of the strided tensor such that `linear_g` is one of the new strided axes.

Yes.

> That is, it's possible to represent the index for `linear_g` as a combination of the indices for `domain`.

Isn't this always possible regardless of if linear or not?

Reply:
added something to `utils.h`.

Reply:
> Isn't this always possible regardless of if linear or not?

Not as a linear combination with integer coefficients, right?

Reply:
> Not as a linear combination with integer coefficients, right?

This isn't the case for linear as well. For example, if `linear_g` is `I0`, and `domain` is `[I0*4]`, assume the index of `I0*4` is `i`, then the index of `I0` is `i / 4`, which is not a linear function of `i`. I think it is more important that `i` is a linear function of `i0`, not the opposite.

Reply:
If `linear_g` is `I0` and `domain` is `[I0/4, 4]` a divisible split, then would we also consider `linear_g` to be linear in `domain`? Or similarly, if `linear_g` is `I0*4` and `domain` is `[I0, 4]`?

Reply:
I think for both case, yes.